Update Components

Lenindelrionicaragua opened the pull request.

tlorent commented on week1/project/src/App.test.js:
```diff
@@ -0,0 +1,8 @@
+import { render, screen } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
+  render(<App />);
+  const linkElement = screen.getByText(/learn react/i);
+  expect(linkElement).toBeInTheDocument();
+});
```
I'm not sure this test is still relevant? 😄

Lenindelrionicaragua replied:
No, it's not being used at all. I apologize for the oversight. I will delete this, along with other automatically generated files.

tlorent commented on week1/project/src/App.js:
```diff
@@ -0,0 +1,62 @@
+  );
+
+  // Handle category click event
+  const handleCategoryClick = (category) => {
+    setSelectedCategory(category);
+    setSelectedProduct(null); // Reset selected product when changing categories
+  };
+
+  // Handle product click event
+  const handleProductClick = (product) => {
+    setSelectedProduct(product);
+    setSelectedCategory(null); // Reset selected category when clicking on a product
+  };
```
Nice job on making sure the categories and products are reset when you click on another category or product 🙌

tlorent commented on week1/project/src/components/CategoryList/CategoryList.js:
```diff
@@ -0,0 +1,42 @@
+  };
+
+  return (
+    <div>
```
If this `div` does not have any stylistic reasons to be here (meaning that it needs to be here because of some CSS, like it's a child of a flex parent) then you can also use a "React Fragment". This will make sure you don't render unnecessary `div`s on the page 😉 

```suggestion
    <>your categories code here</>
```

Lenindelrionicaragua replied:
Okay, I didn't know about this option of avoiding the <div> if it's not needed for stylistic reasons. I will implement this change.

tlorent commented on week1/project/src/components/CategoryList/CategoryList.js:
```diff
@@ -0,0 +1,42 @@
+      <ul className="category-list">
+        {categories.map((category) => (
+          <CategoryListItem
+            key={category}
```
Nice job adding the `key`! 😄

tlorent commented on week1/project/src/components/CategoryList/CategoryListItem.js:
```diff
@@ -0,0 +1,13 @@
+  <li
+    className={isSelected ? "active" : ""}
+    onClick={onClick}
+    style={{ cursor: "pointer" }}
```
I would suggest to avoid using inline styles and add a classname for this. I see that you have used the "styles at a distance pattern" in **CategoryList.css** by targeting elements specifically, but it's recommended that all styles on an element should be caused by class names on that element itself 😄

tlorent commented on week1/project/src/components/ProductList/ProductList.js:
```diff
@@ -0,0 +1,35 @@
+            >
+              <img
+                src={product.image}
+                alt={product.title}
```
Great that you added the `alt` tag! 🙌 It's super important for a11y.

tlorent commented on week1/project/src/components/ProductList/ProductList.js:
```diff
@@ -0,0 +1,35 @@
+import React from "react";
+
+const ProductList = ({ products, selectedCategory, onProductClick }) => {
+  const filteredProducts = Array.isArray(products)
+    ? selectedCategory
+      ? products.filter((product) => product.category === selectedCategory)
+      : products
+    : [];
```
🤩

tlorent replied:
A small tip (something you do not have to change now): you can also destructure the `category`:

```suggestion
  const filteredProducts = Array.isArray(products)
    ? selectedCategory
      ? products.filter(({ category ) => category === selectedCategory)
      : products
    : [];
```

Lenindelrionicaragua replied:
Thank you for pointing out this improvement to make the code more concise. I will apply it. I wasn't aware that I could directly request the property I am interested in for each product, instead of passing each object with all its properties.

tlorent reviewed (commented):
@Lenindelrionicaragua Awesome work! 🙌 I'm impressed by your code, it looks clean and readable and you are already applying great React concepts that you will need in every project.

I do have a few comments, please check them 😄 

Btw, I think that you already did some work that was supposed to be done in week 2? Where the products are clickable. Also, the design does not completely match with the [example version:](https://hyf-react-w1-example.netlify.app) the breakpoints, the categories filter. Can you please have a look at that? 😄

Lenindelrionicaragua commented:
I @tlorent hope this message finds you well. I wanted to inform you that I have implemented the changes you requested in the GitHub repository. I realized that I deviated from the assignment requirements out of curiosity, and I appreciate your guidance to refocus on meeting the specific project requirements.

Here are the key updates I made:

Removed automatically generated files in React, including the test files.
Replaced the use of <div> with <> where unnecessary for styling reasons.
Moved inline styles to the dedicated CSS file for better organization.
Implemented the use of the alt attribute for accessibility (a11y) in images.
Adjusted the product file to directly use the "Category" property instead of filtering each object with all its properties.
Eliminated the "All Categories" option used to display all products, aligning with the example version.
Fine-tuned the letter colors and image sizes to match the design example more closely.
Modified the breakpoints to better align with the example and adjusted the behavior to display products in a single column only on mobile-sized screens.
I also addressed several CSS issues that needed correction. Your observations were insightful and immensely valuable.

Thank you for your guidance, and I appreciate the opportunity to enhance my skills through your feedback.

tlorent commented:
@Lenindelrionicaragua Great work with editing your 🙌 And thanks for the quick update. I just have two comments left: if you look at the [example version](https://hyf-react-w1-example.netlify.app) the "biggest" breakpoint is when you see 4 items next to each other (see screenshot). Also, if you click on the category in the example version then the category is removed and you see _all products_ again. Can you see if you can still fix that? 😄 If it's already in the homework for week 2 then that is also fine.

![Screenshot 2023-12-18 at 10 09 22](https://github.com/HackYourAssignment/React-class45/assets/23385221/7eebbeac-613d-4e6a-a7f2-946caa559820)

Lenindelrionicaragua commented:
@tlorent hi, goodmorning!

Regarding the display of items in 4 columns, it's odd because I only see 3 columns on my end in full-screen mode. However, I've made adjustments to make it display in 4 columns. I'm attaching a screenshot showing how the example looks on my laptop. (How can I prevent these types of errors in the future?)

Regarding the change in the categories section, I have already added it. In fact, I had implemented it for the second-week assignment. :)

![Captura de Pantalla 2023-12-18 a la(s) 11 04 24](https://github.com/HackYourAssignment/React-class45/assets/142052112/130f2e16-221a-45cd-b9f6-bdbee7741d2f)

tlorent commented:
@Lenindelrionicaragua There's still something interesting going on with the breakpoints (see screenshot), but let's continue to week 2 of the homework! The filtering works good now 🙌 

<img width="806" alt="Screenshot 2023-12-21 at 13 39 27" src="https://github.com/HackYourAssignment/React-class45/assets/23385221/2c263152-c27b-4de4-b19e-9db44832bb53">
